packages/openfeature-js/src/lib/client.ts

```diff
@@ -1,106 +1,213 @@
+          valuePromise = provider.getNumberValue(
+            flagId,
+            defaultValue as number,
+            context,
+            options
+          );
           break;
         }
         case 'json': {
-          valuePromise = provider.getObjectValue(id, defaultValue as object, context);
+          valuePromise = provider.getObjectValue(
+            flagId,
+            defaultValue as object,
+            context,
+            options
+          );
           break;
         }
       }
 
       const value = await valuePromise;
-      span.setAttribute(SpanProperties.FEATURE_FLAG_VALUE, value.toString());
-      return value as T;
+      return this.afterEvaluation(allHooks, hookContext, value) as T;
     } catch (err) {
-      console.error(err);
-      span.setAttribute(SpanProperties.FEATURE_FLAG_VALUE, defaultValue.toString());
+      if (this.isError(err)) {
+        this.errorEvaluation(allHooks, hookContext, err);
+      }
       return defaultValue;
     } finally {
-      span.end();
+      this.finallyEvaluation(allHooks, hookContext);
     }
   }
 
-  private startSpan(spanName: string): Span {
-    return this._trace.startSpan(spanName, {
-      attributes: {
-        [SpanProperties.FEATURE_FLAG_CLIENT_NAME]: this._name,
-        [SpanProperties.FEATURE_FLAG_CLIENT_VERSION]:
-          this._version ?? 'unknown',
-        [SpanProperties.FEATURE_FLAG_SERVICE]: this.getProvider().name,
+  private beforeEvaluation(allHooks: Hook[], hookContext: HookContext): HookContext {
+    const mergedContext = allHooks.reduce(
+      (accumulated: Context, hook: Hook): Context => {
+        if (typeof hook?.before === 'function') {
+          return {
+            ...accumulated,
+            ...hook.before(hookContext),
+          };
+        }
+        return accumulated;
       },
+      hookContext.context
+    );
+    hookContext.context = mergedContext;
+    return hookContext;
+  }
+
+  private afterEvaluation(allHooks: Hook[], hookContext: HookContext, flagValue: FlagValue): FlagValue {
+    return allHooks.reduce((accumulated: FlagValue, hook) => {
+      if (typeof hook?.after === 'function') {
+        return hook.after(hookContext, flagValue);
+      }
+      return accumulated;
+    }, flagValue);
+  }
+
+  private finallyEvaluation(allHooks: Hook[], hookContext: HookContext): void {
+    allHooks.forEach((hook) => {
+      if (typeof hook?.finally === 'function') {
+        return hook.finally(hookContext);
+      }
+    });
+  }
+
+  private errorEvaluation(allHooks: Hook[], hookContext: HookContext, err: Error): void {
+    // Workaround for error scoping issue
+    const error = err;
+    allHooks.forEach((hook) => {
+      if (typeof hook?.error === 'function') {
+        return hook.error(hookContext, error);
+      }
     });
   }
 
   private getProvider(): FeatureProvider {
     return this.api.getProvider() ?? NOOP_FEATURE_PROVIDER;
   }
+
+  private isError(err: unknown): err is Error {
```

Comment:
Is this used?

Reply:
Yes, line 151 - I think you wrote this method